CORE-2971 Add attributes to track actual task state transition dates

Smotko opened the pull request.

zidarsk8 commented on src/ggrc_workflows/__init__.py:
```diff
@@ -671,7 +671,16 @@ def handle_cycle_status_change(sender, obj=None, new_status=None,
+      db.session.add(obj)
+    if new_status == 'Finished':
+      obj.finished_date = datetime.now()
+      db.session.add(obj)
```
is this session add needed? The obj should already be in the session if it's being modified

Smotko replied:
I'm not sure. All Signal handlers in workflows __init__.py add objects they are editing to the session...

zidarsk8 replied:
Okay, We can leave it "just to be safe" but generally it's should not be there.

zidarsk8 commented on src/ggrc_workflows/migrations/versions/20160104135243_13e52f6a9deb_add_finished_verified_dates_to_cycle_.py:
```diff
@@ -0,0 +1,38 @@
+  """)
+  op.execute("""
+      UPDATE cycle_task_group_object_tasks
+      SET verified_date = updated_at
```
If a task is verified, then it must have been finished right? In the UI you can't make a task with verified date but without finished date. so I think this should also update both.

Smotko replied:
Confirmed with prasanna. I'll make it so finished date is set as well.

zidarsk8 commented:
Just a note: you have a bad branch name. Should be feature/.

zidarsk8 commented on src/ggrc_workflows/__init__.py:
```diff
@@ -671,7 +671,16 @@ def handle_cycle_status_change(sender, obj=None, new_status=None,
+      obj.verified_date = datetime.now()
+      db.session.add(obj)
+    if new_status == 'Finished':
+      obj.finished_date = datetime.now()
```
You should also set the verified date here to None, since undoing a verification will leave verified date, and the object will be in finished state, which is weird.

Same goes for undoing ifnished. If the state is anything less then finished, these two dates should be set to None.

Smotko commented:
I've fixed the migration, but I've also changed how the logic works after some quick feedback from prasanna. Now after a task is declined the finished date is removed.

zidarsk8 commented on src/ggrc_workflows/__init__.py:
```diff
@@ -671,7 +671,18 @@ def handle_cycle_status_change(sender, obj=None, new_status=None,
+    if new_status == 'Verified':
+      obj.verified_date = datetime.now()
+    elif new_status == 'Finished':
+      obj.finished_date = datetime.now()
```
This is still missing the unset of verified date. 
![capture 2016-01-01 at 20 36 25](https://cloud.githubusercontent.com/assets/566311/12098486/ece00066-b322-11e5-83d0-f99082453746.png)

uskudnik commented on src/ggrc_workflows/__init__.py:
```diff
@@ -671,7 +671,19 @@ def handle_cycle_status_change(sender, obj=None, new_status=None,
 
-# FIXME: Duplicates `ggrc_basic_permissions._get_or_create_personal_context`
+
+@Signals.status_change.connect_via(models.CycleTaskGroupObjectTask)
```
`ggrc_workflows/__init__.py` is getting over 1000LOC, maybe have this in the model? Though unless we make a bigger refactor, that could be more confusing than having it here...

zidarsk8 replied:
Refactor will come in PJ, but we don't wan't to do that for Zucchini.

uskudnik replied:
Yeah, agreed :+1:
